src/lua/init.lua

```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
     return loadfile(file)
 end
 
-local function search_cwd_lib(name)
-    local path = "./?."..soext
-    return package.searchpath(name, path)
-end
+local function merge_strings(paths, templates)
+    local searchpaths = {}
+    paths = type(paths) == 'table' and paths or { paths }
+    templates = templates or {}
 
-local function search_cwd_lua(name)
-    local path = "./?.lua;./?/init.lua"
-    return package.searchpath(name, path)
-end
+    if #templates == 0 then
+        return paths
+    end
+
+    if #paths == 0 then
+        return templates
+    end
 
-local function traverse_rocks(name, pathes_search)
-    local cwd = fio.cwd()
-    local index = string.len(cwd) + 1
-    local strerr = ""
-    while index ~= nil do
-        cwd = string.sub(cwd, 1, index - 1)
-        for i, path in ipairs(pathes_search) do
-            local file, err = package.searchpath(name, cwd .. path)
-            if err == nil then
-                return file
-            end
-            strerr = strerr .. err
+    for _, path in ipairs(paths) do
+        for _, template in pairs(templates) do
+            table.insert(searchpaths, fio.pathjoin(path, template))
         end
-        index = string.find(cwd, "/[^/]*$")
     end
-    return nil, strerr
+
+    return searchpaths
 end
 
-local function search_rocks_lua(name)
-    local pathes_search = {
-        "/.rocks/share/tarantool/?.lua;",
-        "/.rocks/share/tarantool/?/init.lua;",
-    }
-    return traverse_rocks(name, pathes_search)
+local LIB_TEMPLATES = { '?.'..soext }
+local LUA_TEMPLATES = { '?.lua', '?/init.lua' }
+local ROCKS_LIB_TEMPLATES = { ROCKS_LIB_PATH .. '/?.'..soext }
+local ROCKS_LUA_TEMPLATES = { ROCKS_LUA_PATH .. '/?.lua', ROCKS_LUA_PATH .. '/?/init.lua' }
+
+local function search_lib(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LIB_TEMPLATES), ';'))
 end
 
-local function search_rocks_lib(name)
-    local pathes_search = {
-        "/.rocks/lib/tarantool/?."..soext
-    }
-    return traverse_rocks(name, pathes_search)
+local function search_lua(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LUA_TEMPLATES), ';'))
+end
+
+local function traverse_path(path)
+    path = path:sub(1, 1) == '/' and path or fio.abspath(path)
```

Comment:
I've already pointed on always calling abspath, but it changes behavior. We can discuss it once again.

Reply:
Sorry, missed your comment. This should be commented at least, it is not obvoius point.

Reply:
This is undocumented `feature`. Maybe we should just fix it and mention about it in changelog. @Totktonada do you mind?

Reply:
I'll cite Yaroslav's example to bring needed context:

```lua
> traverse_path('/a/b/c/../../A')
---
- - /a/b/c/../../A
  - /a/b/c/../..
  - /a/b/c/..
  - /a/b/c
  - /a/b
  - /a
  - /
...
```

A path with `/../` in a middle could not appear in `traverse_rocks()`, because `fio.cwd()` does not produce such paths. So there is no need to introduce such strange behaviour and so it worth to always call `fio.abspath()`. Ilya, do not I missed something?

Reply:
Ouch, no, `fio.abspath()` doing this work in `setsearchroot()`, so there is no need in `fio.abspath()` here at all.

Reply:
`traverse_path` function now uses `fio.abspath`